[enhancement] Add way to specify global defaults

nepython opened the pull request:
<!--
Before submitting a Pull Request, please make sure you have read
the OpenWISP Contributing Guidelines:
http://openwisp.io/docs/developer/contributing.html#how-to-commit-your-changes-properly
-->

Checks:

- [x] I have manually tested the proposed changes
- [x] I have written new test cases to avoid regressions (if necessary)
- [x] I have updated the documentation (e.g. README.rst)

Closes #106

nepython commented:
The build is failing after I rebased it to the latest changes on `dev` branch. The last commit on `dev` https://github.com/openwisp/openwisp-monitoring/commits/dev too was [failing with the same assertion error](https://travis-ci.org/github/openwisp/openwisp-monitoring/jobs/715362589) (related to changing `http` --> `https`). I am checking what's wrong.

nepython commented on README.rst:
```diff
@@ -1083,7 +1179,7 @@ If you want to extend this view, you will have to perform the additional steps b
         pass
 
-**Step 2: remove the following line from your root ``urls.py`` file:**
+Step 2: remove the following line from your root ``urls.py`` file:
```
**Note**: This is a very small refactoring in `Readme`, not related to this PR. Same for below change.

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,512 @@
+                    '[{site.name}] {notification.target} is not reachable'
+                ),
+                'message': _(
+                    'The device [{notification.target}] {notification.verb} anymore by our ping '
```
Watch out, the markdown link has been removed and only the square brackets have remained:

![Screenshot from 2020-08-07 21-11-40](https://user-images.githubusercontent.com/841044/89700362-2d7c6100-d8f3-11ea-94e8-a41d1c6ee62f.png)

The device should be linked as the word "default" in the other sample notifications.
Applies to the other alerts as well.

If in doubt, ask @pandafy.

nepython replied:
This has been fixed now.

nepython commented on openwisp_monitoring/monitoring/migrations/0020_make_alertsettings_fields_null.py:
```diff
@@ -0,0 +1,32 @@
+    if the values match default values provided in the metric configuration
+    then those values will be set as None
+    """
+    AlertSettings = apps.get_model('monitoring', 'AlertSettings')
```
Thanks, the migration functions now won't make use of `AlertSettings`'s methods.

nepython commented on openwisp_monitoring/monitoring/admin.py:
```diff
@@ -9,8 +10,21 @@
+            kwargs['initial'] = {
+                'tolerance': instance.get_tolerance,
+                'operator': instance.get_operator,
+                'threshold': instance.get_threshold,
```
Yep, I too noticed this yesterday night and was working on `full_clean`. Added a regression test for the same,
https://github.com/openwisp/openwisp-monitoring/blob/3f8215fc982161344eea40b390bb17934188595d/openwisp_monitoring/monitoring/tests/test_models.py#L254-L272

nepython commented:
So currently 3 things need to be done:
* [ ] Making the **docs** more friendly by giving more details and reordering few alike topics after some discussion.
* [ ] Changing **tolerance field** input from `seconds` to `minutes`.
* [ ] Fixing the percentage display in **notifications** (it should display `70%` and not `0.7`)

nepython commented on openwisp_monitoring/monitoring/migrations/0020_make_alertsettings_fields_null.py:
```diff
@@ -0,0 +1,32 @@
+            alert_settings.tolerance = None
+        if alert_settings.operator == alert_setting.config_dict['operator']:
+            alert_settings.operator = None
+        alert_settings.save()
```
I have added backward migrations but please note that there is a conversion of `seconds` --> `minutes` (for **tolerance** field) which will show a minor change. So, say if you had `seconds=61` for one `AlertSettings`  before migrating. After migrating and then running reverse migrations you will have `seconds=60`.

nemesisdesign commented on openwisp_monitoring/monitoring/base/models.py:
```diff
@@ -482,26 +483,68 @@ class AbstractAlertSettings(TimeStampedEditableModel):
+        return self.tolerance or self.config_dict['tolerance']
+
+    @property
+    def get_operator(self):
```
methods named `get_` should not be decorated with `@property`.

I see mainly two possibilities:

- either leave these methods undecorated and call them like `instance.get_threshold()`
- rename the model fields as `custom_<field_name>` and rename these decorated methods to remove `get_` 

eg:

```python
@property
def tolerance(self):
    return self.custom_tolerance or self.config_dict['tolerance']
```

nepython commented:
> **Question**: what if a user wants to change the default tolerance of the ping metric?
Do they have to redefine everything or can we find a way to allow them to only specify the differences?

Well for `AlertSettings` specifically we have `custom_threshold`, `custom_operator`, `custom_tolerance` but I get your query, what if the user wants to change say `notification`, right? I got you covered, see below. :smiley:
> Maybe we could add a new function specifically for this, eg:
>```python
># sloppy pseudocode alert!!
>
>modify_thing('ping', {'alerts': {'tolerance': 3}}
>```
>It's just an idea, if you have a better idea let me know.

@nemesisdesign, I took inspiration from your work with `charts` and adapted the setting `OPENWISP_MONITORING_CHARTS` for metrics. This should cover your use case and also maintains consistency, I think https://github.com/openwisp/openwisp-monitoring/tree/issues/106-global-defaults#openwisp_monitoring_metrics. I know `README` needs improvement :sweat_smile:

nemesisdesign commented on tests/openwisp2/sample_monitoring/migrations/0001_initial.py:
```diff
@@ -167,19 +171,33 @@ class Migration(migrations.Migration):
+                    'custom_threshold',
+                    models.FloatField(
+                        blank=True,
+                        help_text='sets the threshold for collected metric data to send alerts, you can set threshold value as a float',
```
I would just keep `threshold value`, which will be clear to most users

nemesisdesign commented on tests/openwisp2/sample_monitoring/migrations/0001_initial.py:
```diff
@@ -167,19 +171,33 @@ class Migration(migrations.Migration):
                         default=0,
-                        help_text='for how long should the value be crossed before an alert is sent? The maximum allowed is 604800 seconds (7 days)',
-                        validators=[django.core.validators.MaxValueValidator(604800)],
+                        help_text='for how long should the value be crossed before an alert is sent? The maximum allowed is 10080 minutes (7 days)',
```
let's change this to something more explicit and useful:

```
for how many minutes should the threshold value be crossed before an alert is sent? A value of zero means the alert is sent immediately
```

nemesisdesign commented on README.rst:
```diff
@@ -741,14 +726,165 @@ An example usage is shown below.
+|**metric_configuration**: | A ``dict`` defining configuration of the metric.     |
++--------------------------+------------------------------------------------------+
+
+An example usage has been shown below.
```
Please summarize what the following example does.

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+                ),
+                'message': _(
+                    'The device [{notification.target}]({notification.target_link}) '
+                    '{notification.verb} anymore by our ping messages.'
```
please remove ` anymore by our ping messages`

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+            },
+            'recovery': {
+                'name': 'Ping RECOVERY',
+                'verb': 'has become reachable',
```
let's change this to `is reachable again`

nemesisdesign replied:
make string translatable

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+                'verb': 'has become reachable',
+                'level': 'info',
+                'email_subject': _(
+                    '[{site.name}] {notification.target} is reachable again'
```
here you can use the verb instead of duplicating "is reachable again"

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+                ),
+                'message': _(
+                    'The device [{notification.target}]({notification.target_link}) '
+                    '{notification.verb} again by our ping messages.'
```
let's remove ` again by our ping messages`

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+        'notification': {
+            'problem': {
+                'name': 'Ping PROBLEM',
+                'verb': 'cannot be reached anymore',
```
`is not reachable`

nemesisdesign replied:
make string translatable

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+                'verb': 'cannot be reached anymore',
+                'level': 'warning',
+                'email_subject': _(
+                    '[{site.name}] {notification.target} is not reachable'
```
`[{site.name}] PROBLEM: {notification.target} {notification.verb}`

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+        'alert_settings': {'operator': '<', 'threshold': 1, 'tolerance': 0},
+        'notification': {
+            'problem': {
+                'name': 'Ping PROBLEM',
```
what's this key used for? Is this shown to the user?

nepython replied:
I am not sure how it is used internally, I didn't observe it via the admin. @pandafy might be able to confirm / correct me.

nemesisdesign replied:
@nepython according to the [openwisp-notifications README](https://github.com/openwisp/openwisp-notifications/tree/dev#register-notification-type), this key is not required.
Please double check and remove it if it's not needed.
@pandafy anything to add?

pandafy replied:
@nepython this is what is shown to the user. Please rename it to `verbose_name`. It was set incorrectly earlier. :sweat_smile: 
I fixed it in [#189](https://github.com/openwisp/openwisp-monitoring/pull/189/files#diff-583ce6230e4e945550ab510f1b8be8d8R154), but please make such changes here also. It will be used while [displaying notification settings](https://user-images.githubusercontent.com/32094433/89739293-2b560600-da9d-11ea-9bdb-ce92c89dc3be.gif).

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+                'verb': 'is not applied',
+                'level': 'warning',
+                'email_subject': _(
+                    '[{site.name}] {notification.target} has problem in configuration'
```
missing PROBLEM

it should say something like: 

PROBLEM device X configuration status issue

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+                ),
+                'message': _(
+                    'The configuration for device [{notification.target}]'
+                    '({notification.target_link}) {notification.verb} properly.'
```
The configuration **of** device X ......... has not been applied in a timely manner.

nemesisdesign commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -0,0 +1,519 @@
+                ),
+                'message': _(
+                    'The device [{notification.target}]({notification.target_link}) '
+                    '{notification.verb} successfully.'
```
has been applied again successfully

nepython commented:
Addressed all reviews /  issues besides docs (working on it).

> Why did you remove RECOVERY: and PROBLEM: from the email subjects?
Please add these back.
Every email alert must have PROBLEM or RECOVERY stated clearly in its subject

The issue had 3 notification examples. None had `RECOVERY` or `PROBLEM` in email subjects, hence I removed them. :sweat_smile:. Reverting.

nemesisdesign commented:
> **PS**: Docs updated as per the review but topics can be made more ordered to improve a new user's experience. Maybe we can do this in #70 since it's a dedicated issue for improving `README`.

As a general rule, it's better to work on documentation while working on a specific feature, because our minds are already focused on it and therefore it requires less effort compared to postponing it.

nemesisdesign commented on openwisp_monitoring/monitoring/tests/__init__.py:
```diff
@@ -7,14 +7,66 @@
+
+test_notification = {
+    'problem': {
+        'name': 'Monitoring Alert',
```
should be `verbose_name`, right?

nemesisdesign commented on openwisp_monitoring/monitoring/tests/__init__.py:
```diff
@@ -7,14 +7,66 @@
+        'message': default_message,
+    },
+    'recovery': {
+        'name': 'Monitoring Alert',
```
should be `verbose_name`, right?

nemesisdesign commented on README.rst:
```diff
@@ -618,71 +709,173 @@ In case you just want to change the colors used in a chart here's how to do it:
+        'alert_settings': {'operator': '<', 'threshold': 1, 'tolerance': 0},
+        'notification': {
+            'problem': {
+                'name': 'Ping PROBLEM',
```
should be `verbose_name`, right?

nemesisdesign commented on README.rst:
```diff
@@ -618,71 +709,173 @@ In case you just want to change the colors used in a chart here's how to do it:
+                ),
+            },
+            'recovery': {
+                'name': 'Ping RECOVERY',
```
should be `verbose_name`, right?

nemesisdesign commented on README.rst:
```diff
@@ -547,6 +532,112 @@ network topology module will trigger the monitoring checks.
+        'config_applied': {
+            'notification': {
+                'problem': {
+                    'verb': _('has not been applied'),
```
does it have any effect changing `verbose_name` here? If it does, please include it in the example, eg: "PING problem changed name" or something like that.

nepython replied:
Yes, it will have. Earlier I wasn't very sure how helpful it might be earlier. It should be after `Notification Settings`, I think. Added the `verbose_name` as `Configuration PROBLEM` (default is `Configuration Applied PROBLEM` so just shortened it).

nemesisdesign commented on README.rst:
```diff
@@ -547,6 +532,112 @@ network topology module will trigger the monitoring checks.
+                    ),
+                },
+                'recovery': {
+                    'verb': _('configuration has been applied again'),
```
same here
